spec/mempool/gossip/dog.md

```diff
@@ -0,0 +1,606 @@
+            rc' = rc.put(node, _rc1),
+            dr' = dr,
+        }
+    ```
+    where `replyHaveTx` will send a `HaveTx` message if `tx` comes from a peer and `HaveTx` messages are
+    not blocked:
+    ```bluespec "actions" +=
+    pure def replyHaveTx(node, _incomingMsgs, _rc, optionalSender, tx) =
+        if (optionalSender.isSome() and not(_rc.isHaveTxBlocked))
+            val targets = optionalSender.optionToSet()
+            (node.multiSend(_incomingMsgs, targets, HaveTxMsg(hash(tx))), _rc.blockHaveTx())
+        else (_incomingMsgs, _rc)
+    ```
+    Immediately after sending the `HaveTx` message the controller will block sending a new one.
+
+### Handling incoming messages
+
+In this subsection we define how to handle each type of message received from a peer (the `sender`).
+```bluespec "actions" +=
+action handleMessage(node, _incomingMsgs, sender, msg) =
+    match msg {
+    | TxMsg(tx) => node.tryAddTx(_incomingMsgs, Some(sender), tx)
+    | HaveTxMsg(txID) => node.handleHaveTxMessage(_incomingMsgs, sender, txID)
+    | ResetMsg => node.handleResetMessage(_incomingMsgs, sender)
+    }
+```
+
+* **Handling `HaveTx` messages**
+
+    Upon receiving `HaveTxMsg(txID)` from `sender`, `node` disables the route `firstSender -> sender`, 
+    where `firstSender` is the first node in `txID`'s list of senders, if `txID` actually
+    comes from a peer. This action will decrease the traffic to `sender`.
+    ```bluespec "actions" +=
+    action handleHaveTxMessage(node, _incomingMsgs, sender, txID) = all {
+        val txSenders = node.sendersOf(txID)
+        dr' = if (length(txSenders) > 0) dr.disableRoute(node, txSenders[0], sender) else dr,
+        incomingMsgs' = _incomingMsgs,
+        peers' = peers,
+        mempool' = mempool,
+        senders' = senders,
+        rc' = rc,
+    }
+    ```
+    The list of `tx`’s senders contains the node IDs from which `node` received the transaction,
+    ordered by the arrival time of the corresponding messages. To avoid disabling the routes from
+    all those senders at once, the protocol picks the first sender in the list, which is the first
+    peer from which `node` received `tx`. Subsequent entries in the list are nodes whose transaction
+    messages arrived later as duplicates. As such, most routes from those peers to `node` will
+    eventually be disabled, with most traffic coming primarily from the first peer.
+
+* **Handling Reset messages**
+
+    Upon receiving `ResetMsg`, `node` removes any route that has `sender` as source or target. 
+    ```bluespec "actions" +=
+    action handleResetMessage(node, _incomingMsgs, sender) = all {
+        dr' = dr.update(node, drs => drs.enableRoute(sender)),
+        incomingMsgs' = _incomingMsgs,
+        peers' = peers,
+        mempool' = mempool,
+        senders' = senders,
+        rc' = rc,
+    }
+    ```
+    This will allow traffic to flow again to the sender, and other nodes will dynamically adapt to the
```

Comment:
To and from the sender, right?

Reply:
Only to the sender. I fixed the spec and updated the text.